### DISCUSSION TITLE
How can I test commands?

### DISCUSSION THREAD

**Discussion by goldingdamien** [evilmartians/lefthook#624] [Q&A] [answered]
I want to be able to test commands for the following:
- "glob" picks up files.
- "run" cli command execution with files(passed manually).
- "root" tested together.

It should work without any form of git staging and committing regardless of the hook tested.
Files should be able to be manually specified regardless of parameter(staged_files, etc.) used.
It should work with commands that invoke npm scripts that contain non-global libraries for execution via name such as { "scripts": { "test": "vitest" } }.

I thought this would be easy to do via lefthook run, but couldn't get it working. Please advise.
https://github.com/evilmartians/lefthook/blob/master/docs/usage.md#lefthook-run

**Comment by EDMPhoenix** [comment]


**Reply by mrexox** [reply, marked as answer]
Thank you @EDMPhoenix! I just want to add that `lefthook` supports a few extra options for the `lefthook run <hook>` command. 

Let's consider the following config:

```yml
pre-commit:
  commands:
    lint:
      run: yarn lint {staged_files}
      glob: "*.ts"
      root: "frontend/"
```

When you run `lefthook run pre-commit` the `lint` command will probably be skipped because there're no staged files matching the `glob` under the `root`. Instead you can do the following:

```bash
lefthook run pre-commit --force --commands lint --file frontend/index.js --file backend/main.go
```

Or if you want to pass all your files you can do this:

```bash
find . -type f -print0 | lefthook run pre-commit --force --commands lint --files-from-stdin
```

See more in `lefthook run -h`
